### PR TITLE
Fix issues detected by psalm on 2.7

### PR DESCRIPTION
**Pull request by orklah**
This PR is basically #7986 but targetted on 2.8 as suggested.

I'll copy the text from there:

> Hi,
> 
> This PR is the result of fixing easy issues from the two first levels of Psalm.
> 
> This is the starting point of trying to begin #7762. If this goes smoothly, my goal is to continue fixing errors by slowly increasing the Psalm's levels. Eventually, the idea would be to add Psalm to the CI and to add template annotation to interfaces and exposed classes to help tools like phpstan and psalm understand Doctrine without the need of stubs or plugins.
> 
> I'm always trying to learn so don't hesitate to ask for changes or anything!
> 
> Thanks!

I didn't expect such difference between master and 2.8! Great job guys, I can't wait for future versions!

It seems that the CI is broken on 2.8. I tried to run phpcbf locally but there are massive changes, I didn't dare apply them here, so there may be some CS errors. If the CI is fixed and the errors are still there in the future, I'll correct them.

I'll close #7986, some changes there were specific to the master branch due to code movements but let's fix that later.

### CODE REVIEW

**Comment by greg0ire** [comment]
Restarting the build

File: lib/Doctrine/ORM/Configuration.php
Comment (greg0ire) [inline lib/Doctrine/ORM/Configuration.php]: That kind of trailing whitespace makes the build fail. Please run `vendor/bin/phpcbf` to fix that

**Comment by orklah** [comment]
Issues in Scrutinizer are correct but were not introduced by this PR. Can this be merged as is?

File: lib/Doctrine/ORM/Configuration.php
Comment (greg0ire) [inline lib/Doctrine/ORM/Configuration.php]: When I see https://github.com/doctrine/data-fixtures/pull/332, I'm not sure we should use `psalm-return`… I think we could just use return and remove line 443… WDYT? Apparently it's recognized by recent versions of phpstan as well: https://github.com/doctrine/data-fixtures/issues/337

**Comment by greg0ire** [comment]
Yeah don't worry about Scrutinizer

File: lib/Doctrine/ORM/Configuration.php
Comment (orklah) [reply lib/Doctrine/ORM/Configuration.php]: Yeah, Phpstan reads most psalm annotations and understand almost everything psalm uses. I kept the return annotation mainly for other tools like PHPStorm who don't understand it yet. For example, when I added types in https://github.com/Roave/BetterReflection, Ocramius prefered to keep standard notations clean as long as psalm's and phpstan's 
notations are not understood widely: https://github.com/Roave/BetterReflection/pull/531#pullrequestreview-342848231

For `class-string`, it may seem a trivial difference in PHPStorm (string vs mixed), but when we will replace things like `RandomClass[]` by their more detailed `list<RandomClass>`, PHPStorm won't even read it as an array, and autocompletion will be broken. Depending on the IDEs used by maintainers, it can be a big drawback

**Review by greg0ire — DISMISSED** [review]

**Review by SenseException — DISMISSED** [review]
Scrutinizer needs a restart sometimes and it gets back to be green again.

**Comment by orklah** [comment]
Thanks for approving. What's the next step on this? Should we wait for more approval or is there something to do?

Thanks

**Comment by greg0ire** [comment]
More approvals would be great indeed, especially from people who are familiar with the annotations (@Ocramius maybe?) @SenseException I don't know how familiar you are with psalm annotations and how we use them inside Doctrine, feel free to merge.

Meanwhile, please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already push. Don't forget the `--force` option otherwise git will try to merge both things together.

**Comment by SenseException** [comment]
Not familiar enough to merge it without hesitation.

**Comment by greg0ire** [comment]
I've become more familiar with Psalm now. I think this could be merged, but should be merged on 2.7, since it is just code comments, and in order to keep differences between branches minimal.

**Comment by orklah** [comment]
@greg0ire great! I don't really know how to switch branches though, do you also have an explanation like the one for rebase? (it was extremely useful btw, I learned to rebase thanks to that explanation and it made me understand a lot about git)

**Comment by greg0ire** [comment]
Start by clicking the edit button and changing the target branch to 2.7. You should see a lot of commit appear. What you would want to rebase is only the commit you added on top of 2.8.x, so that would be `git rebase --onto origin/2.7 2.8.x psalm-types-2.8`. After you can force push and things should be back to normal.

Another way to do the same thing would be this:


1. Fetch all new commits: `git fetch --all`. ⬇️ 
2. Rebase on what you fetched interactively: ✂️ 
    1. `git rebase -i origin/2.7`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin repository.
    2. A window will show up with many lines, remove every line but the last one, which should correspond to your commit.
   2. If you run into a conflict: 💥 
        1. fix it with `git mergetool`. 😷 
        1. continue on your merry way with `git rebase --continue`. ⏩ 
3. Force push to overwrite all this : `git push --force`. ⬆️

**Comment by orklah** [comment]
Mmh, that went pretty smoothly, it's suspicious!

```C:\Users\User\PhpstormProjects\orm>git fetch --all
Fetching origin
Fetching upstream
remote: Enumerating objects: 547, done.
remote: Counting objects: 100% (547/547), done.
remote: Compressing objects: 100% (31/31), done.
remote: Total 834 (delta 511), reused 532 (delta 510), pack-reused 287
Receiving objects: 100% (834/834), 438.74 KiB | 1.17 MiB/s, done.
Resolving deltas: 100% (596/596), completed with 291 local objects.
From https://github.com/doctrine/orm
   fdad48278..6780a963f  2.7             -> upstream/2.7
   a236a83fa..850d57e79  2.8.x           -> upstream/2.8.x
   e77091399..9a8e1fb47  master          -> upstream/master
 * [new branch]          new-embeddables -> upstream/new-embeddables
 * [new tag]             v2.7.1          -> v2.7.1
 * [new tag]             v2.7.2          -> v2.7.2

C:\Users\User\PhpstormProjects\orm>git rebase -i upstream/2.7
error: cannot rebase: You have unstaged changes.
error: Please commit or stash them.

C:\Users\User\PhpstormProjects\orm>git rebase -i upstream/2.7
Successfully rebased and updated refs/heads/psalm-types-2.8.

C:\Users\User\PhpstormProjects\orm>git push --force
Enumerating objects: 34, done.
Counting objects: 100% (34/34), done.
Delta compression using up to 8 threads
Compressing objects: 100% (16/16), done.
Writing objects: 100% (18/18), 2.15 KiB | 2.15 MiB/s, done.
Total 18 (delta 15), reused 0 (delta 0)
remote: Resolving deltas: 100% (15/15), completed with 14 local objects.
To https://github.com/orklah/orm.git
 + 38eb578b1...6e6c28986 psalm-types-2.8 -> psalm-types-2.8 (forced update)
```

I wanted to try the first solution but I messed up and used the second one without noticing. but it finished successfully but I still see `orklah wants to merge 1 commit into doctrine:2.8.x from orklah:psalm-types-2.8`

So I'm not sure if everything went well. Can you check if it's okay please?

I noticed a new error on psalm, it is indeed an issue in code (we try to instantiate a potential null string). I'd prefer to avoid touching code unless I'm more comfortable with the repo. If it's okay with you, I'll just add it to the baseline or add an issueHandler in psalm.xml

**Comment by greg0ire** [comment]
> but I still see orklah wants to merge 1 commit into doctrine:2.8.x from orklah:psalm-types-2.8 on the github PR

That's expected, isn't it? That's what I expected at least.

**Comment by greg0ire** [comment]
> I'd prefer to avoid touching code unless I'm more comfortable with the repo. If it's okay with you, I'll just add it to the baseline or add an issueHandler in psalm.xml

No please don't, I will look into it for you.

**Comment by orklah** [comment]
> That's expected, isn't it? That's what I expected at least.

Well, wasn't the whole point to switch from 2.8 to 2.7? The fact that it still displays 2.8 anywhere that isn't the name of my branche seems disturbing.

> No please don't, I will look into it for you.

Great, thanks!

**Comment by greg0ire** [comment]
> Well, wasn't the whole point to switch from 2.8 to 2.7? The fact that it still displays 2.8 anywhere that isn't the name of my branche seems disturbing.

Oh right you did not switch the head branch with the edit button

**Comment by greg0ire** [comment]
Can you please allow pushes from maintainers? Should be a checkbox, bottom right

**Comment by orklah** [comment]
Thanks for the change, I didn't know you had to change it there too.

The checkbox was already checked. I unchecked it and then rechecked just to be sure.

**Comment by greg0ire** [comment]
Well, for some reason I have this:

```
gpf orklah HEAD:psalm-types --no-verify
Enumerating objects: 36, done.
Counting objects: 100% (36/36), done.
Delta compression using up to 4 threads
Compressing objects: 100% (17/17), done.
Writing objects: 100% (19/19), 2.94 KiB | 2.94 MiB/s, done.
Total 19 (delta 16), reused 0 (delta 0)
remote: Resolving deltas: 100% (16/16), completed with 15 local objects.
To github.com:orklah/orm.git
 ! [remote rejected]     HEAD -> psalm-types (permission denied)
error: failed to push some refs to 'git@github.com:orklah/orm.git'
```

Not sure if related to this: https://twitter.com/nicolasgrekas/status/1258709224734982146

**Comment by greg0ire** [comment]
Here are my changes BTW:

```
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
modified: lib/Doctrine/ORM/Query/FilterCollection.php
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@ lib/Doctrine/ORM/Query/FilterCollection.php:23 @
namespace Doctrine\ORM\Query;

use Doctrine\ORM\EntityManagerInterface;
use function assert;

/**
 * Collection class for all the query filters.
@ lib/Doctrine/ORM/Query/FilterCollection.php:114 @ class FilterCollection
        if ( ! $this->isEnabled($name)) {
            $filterClass = $this->config->getFilterClassName($name);

            assert($filterClass !== null);
 
            $this->enabledFilters[$name] = new $filterClass($this->em);

            // Keep the enabled filters sorted for the hash
```

**Comment by greg0ire** [comment]
You forgot the `use` statement. Please squash all of your commits together, we should try not to have commits that fail the build.

**Comment by orklah** [comment]
I'm really bad with CS rules :D I should run them locally! It should be ok now

**Review by greg0ire — DISMISSED** [review]

**Comment by greg0ire** [comment]
Trying to get a report from Travis, without success it seems :/

**Comment by orklah** [comment]
yeah, don't know what's up with that...

**Comment by greg0ire** [comment]
I have the same issue with this other PR: https://github.com/doctrine/cache/pull/341

**Comment by greg0ire** [comment]
Seems to work better now :)

File: lib/Doctrine/ORM/Query/SqlWalker.php
Comment (greg0ire) [inline lib/Doctrine/ORM/Query/SqlWalker.php:146]: I was going to merge, then saw this… Isn't that equivalent to
```suggestion
     * @psalm-var array<string, array{metadata: ClassMetadata}>
```
Or is there something I am missing?

File: lib/Doctrine/ORM/Query/SqlWalker.php
Comment (greg0ire) [inline lib/Doctrine/ORM/Query/SqlWalker.php:238]: same here

File: lib/Doctrine/ORM/Query/SqlWalker.php
Comment (orklah) [reply lib/Doctrine/ORM/Query/SqlWalker.php:146]: to be honest, I don't remember that part. I didn't use psalter and I don't see why I would have added that manually. It was probably a psalm suggestion, even if that doesn't make a lot of sense. I'll revert that

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks a lot @orklah

**Pull request merged**
